### PR TITLE
Fix gitlab oauth token validation request

### PR DESCRIPTION
**Pull request by vinokurig**
### What does this PR do?
Rework the gitlab oauth token validation check to verify if the test response has data in the returned object.

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/22836

### How to test this PR?

1. Deploy che with the PR image: `quay.io/eclipse/che-server:pr-674`
2. Configure [Gitlab oauth](https://eclipse.dev/che/docs/stable/administration-guide/configuring-oauth-2-for-gitlab/).
3. Start a workspace from a Gitlab repository with a devfile.
4. Go to dashboard -> User Preferences -> Git Services tab, see: the Gitlab provider is in the `authorised` state.
5. Revoke the Gitlab authorisation.

See: the authorisation state of the Gitlab provider has changed to `unauthorised`.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Review by tolusha — APPROVED** [review]

**Comment by dmytro-ndp** [comment]
@vinokurig : hello,
Question about test scenario in description:
> How to test this PR?
> 1. Deploy che with the PR image: `quay.io/eclipse/che-server:pr-`

Did you mean `quay.io/eclipse/che-server:pr-674` ?

**Comment by vinokurig** [comment]
@dmytro-ndp 
>Question about test scenario in description:
>
>>How to test this PR?
>>
>>Deploy che with the PR image: quay.io/eclipse/che-server:pr-
>Did you mean quay.io/eclipse/che-server:pr-674 ?

Right, When I start creating a che-server pull request I don't know the id before it is opened. This time I forgot to update the description after the pr creation. Updated the description.

File: wsmaster/che-core-api-auth-gitlab/src/main/java/org/eclipse/che/security/oauth/GitLabOAuthAuthenticator.java [hunk 1]
```diff
@@ -93,7 +94,8 @@ public OAuthToken getToken(String userId) throws IOException {
       if (token == null
           || token.getToken() == null
           || token.getToken().isEmpty()
-          || getJson(gitlabUserEndpoint, token.getToken(), GitLabUser.class) == null) {
+          || isNullOrEmpty(
```
Comment (ibuziuk) [inline wsmaster/che-core-api-auth-gitlab/src/main/java/org/eclipse/che/security/oauth/GitLabOAuthAuthenticator.java:97]: @vinokurig could you clarify the principle difference and how it fixes gitlab validation? can we add a unit test for that ?

File: wsmaster/che-core-api-auth-gitlab/src/main/java/org/eclipse/che/security/oauth/GitLabOAuthAuthenticator.java [hunk 2]
```diff
@@ -93,7 +94,8 @@ public OAuthToken getToken(String userId) throws IOException {
...
+              getJson(gitlabUserEndpoint, token.getToken(), GitLabUser.class).getId())) {
         return null;
```
Comment (ibuziuk) [inline wsmaster/che-core-api-auth-gitlab/src/main/java/org/eclipse/che/security/oauth/GitLabOAuthAuthenticator.java:99]: returning null is a bad practice in general

**Review by dmytro-ndp — APPROVED** [review]
Test scenario from description with additional steps has tested successfully using `quay.io/eclipse/che-server:pr-674` image using private repo https://gitlab-gitlab-system.apps.git.crw-qe.com/admin-user/private-repo.git with Gitlab Server on OCP 4.14.

_Remark_: Git Services tab showed state "unauthori**z**ed", not "unauthori**s**ed".
I would suggest using the same vocabulary in the test scenario as used in Eclipse Che UI.

Test scenario with additional steps:
1. Deploy che with the PR image: quay.io/eclipse/che-server:pr-674
2. Configure [Gitlab oauth](https://eclipse.dev/che/docs/stable/administration-guide/configuring-oauth-2-for-gitlab/).
3. Start a workspace from a Gitlab repository with a devfile https://gitlab-gitlab-system.apps.git.crw-qe.com/admin-user/private-repo.git.
3.a See page to authorize OAuth on GitLab server during the workspace startup.
4. Go to dashboard -> User Preferences -> Git Services tab, see: the Gitlab provider is in the authorized state.
5. Revoke the Gitlab authorization.
6. See: the authorization state of the Gitlab provider has changed to `unauthorized`.
7. Refresh dashboard page
8. See: the authorization state of the Gitlab provider is `unauthorized`.
9. Start a new workspace from a Gitlab repository with a devfile https://gitlab-gitlab-system.apps.git.crw-qe.com/admin-user/private-repo.git. 
10. Go to VS Code Editor terminal and make sure `git config --list | grep user.name` returns `TestUser`.
11. See page to authorize OAuth on GitLab server during the workspace startup.
12. Start a new workspace from a Gitlab repository with a devfile https://gitlab-gitlab-system.apps.git.crw-qe.com/admin-user/private-repo.git.
13. See NO page to authorize OAuth on GitLab server during the workspace startup.

Test screencast: [screencast-bpconcjcammlapcogcnnelfmaeghhagj-2024.04.03-11_59_12.webm](https://github.com/eclipse-che/che-server/assets/1197777/1388830b-0f3a-43e1-8a9c-077b6fc21f81)

File: wsmaster/che-core-api-auth-gitlab/src/main/java/org/eclipse/che/security/oauth/GitLabOAuthAuthenticator.java (hunk 2 above)
Comment (vinokurig) [reply wsmaster/che-core-api-auth-gitlab/src/main/java/org/eclipse/che/security/oauth/GitLabOAuthAuthenticator.java:95]: Agree, added a `TODO` for refactoring to `Optional`.

File: wsmaster/che-core-api-auth-gitlab/src/main/java/org/eclipse/che/security/oauth/GitLabOAuthAuthenticator.java (hunk 1 above)
Comment (vinokurig) [reply wsmaster/che-core-api-auth-gitlab/src/main/java/org/eclipse/che/security/oauth/GitLabOAuthAuthenticator.java]: > could you clarify the principle difference and how it fixes gitlab validation?

Currently, in the `main` branch  we check that the `GitLabUser` object is not `null`. If we request the GitLab user with an invalid token, we will not get `null`, but a `GitLabUser` object with empty fields. To fix that I changed the check to verify if the `userId` field is not `null`.
> can we add a unit test for that ?

Done

**Review by ibuziuk — APPROVED** [review]

**Pull request merged**